File: Src2/Icm.Core/CommandLine/CommandLineExtensions.vb
Imports System.Runtime.CompilerServices
Imports res = My.Resources.CommandLineTools
Imports Icm.Collections
Imports Icm.ColorConsole

Namespace Icm.CommandLineTools

    Public Module CommandLineExtensions

        ''' <summary>
        '''  Prints the set of instructions for a CommandLine.
        ''' </summary>
        ''' <remarks></remarks>
        <Extension>
        Public Sub Instructions(cmdline As CommandLine)
            Dim o As NamedParameter

            Console.Write("{0}: ", res.S_HELP_USAGE, System.Diagnostics.Process.GetCurrentProcess.ProcessName)

            Using New ColorSetting(ConsoleColor.Cyan)
                Console.Write(System.Diagnostics.Process.GetCurrentProcess.ProcessName)

                For Each param In cmdline.NamedParameters
                    If param.IsRequired Then
                        Console.Write(" --{0}", param.LongName)
                    Else
                        Console.Write(" [--{0}", param.LongName)
                    End If
                    For Each subarg In param.SubArguments
                        If subarg.IsRequired Then
                            Console.Write(" " & subarg.Name)
                        Else
                            Console.Write(" [{0}]", subarg.Name)
                        End If
                    Next
                    If Not param.IsRequired Then
                        Console.Write("]")
                    End If

                Next
            End Using
            Using New ColorSetting(ConsoleColor.Green)

                With cmdline.MainParameters
                    For i = 1 To .Minimum
                        Console.Write(" {0}", .GetArgumentName(i).Name)
                    Next
                    If .Maximum.HasValue Then
                        For i = .Minimum + 1 To .Maximum.V
                            Console.Write(" [{0}]", .GetArgumentName(i).Name)
                        Next
                    Else
                        Console.Write(" {{{0}}}", .UnboundArgumentsName.Name)
                    End If
                End With
            End Using
            Using New ColorSetting(ConsoleColor.White)
                Console.WriteLine()
                Console.WriteLine()
                Console.Write("{0}:", res.S_OPTIONS)
                Console.WriteLine()

                For Each o In cmdline.NamedParameters
                    Dim subargs As New List(Of String)

                    For Each s As SubArgument In o.SubArguments
                        If s.IsRequired Then
                            subargs.Add(s.Name)
                        ElseIf s.IsOptional Then
                            subargs.Add(String.Format("[{0}]", s.Name))
                        ElseIf s.IsList Then
                            subargs.Add(String.Format("{{{0}}}", s.Name))
                        End If
                    Next
                    Using New ColorSetting(ConsoleColor.Cyan)
                        Console.WriteLine("  -{0} {1}", o.ShortName, subargs.JoinStr(" "))
                        Console.WriteLine("  --{0} {1}", o.LongName, subargs.JoinStr(" "))
                    End Using
                    Console.WriteLine("      {0}{1}", If(o.IsRequired, res.S_HELP_REQUIRED, ""), o.Description)
                Next

                Console.WriteLine()
                With cmdline.MainParameters
                    For Each key In .ParamDictionary.Keys
                        Using New ColorSetting(ConsoleColor.Green)
                            Console.WriteLine("  {0}", key)
                        End Using
                        Console.WriteLine("      {0}", .ParamDictionary(key).Description)
                    Next
                End With
                Console.WriteLine()
                Console.WriteLine(res.S_SUBARGUMENTS_SEPARATED)
            End Using

        End Sub

        ''' <summary>
        '''  Prints the set of values for a CommandLine.
        ''' </summary>
        ''' <remarks></remarks>
        <Extension>
        Public Sub Print(cmdline As CommandLine)
            Using New ColorSetting(ConsoleColor.Cyan)
                For Each param In cmdline.NamedParameters
                    If cmdline.IsPresent(param.ShortName) Then
                        ColorConsole.ColorConsole.Write(ConsoleColor.Cyan, "{0}=", param.LongName)
                        ColorConsole.ColorConsole.WriteLine(ConsoleColor.White, "({0})", cmdline.GetValues(param.ShortName).JoinStr(
                                "", """", ",", """", ""))
                    Else
                        ColorConsole.ColorConsole.Write(ConsoleColor.Cyan, "{0}=", param.LongName)
                        ColorConsole.ColorConsole.WriteLine(ConsoleColor.White, "UNDEFINED")
                    End If
                Next
            End Using
            Using New ColorSetting(ConsoleColor.Green)

                Console.WriteLine(cmdline.MainValues.JoinStr(
                            "(", """", ",", """", ")"))
            End Using

            Console.WriteLine()

        End Sub

        <Extension>
        Public Function GetInteger(cmdline As CommandLine, key As String) As Integer?
            If cmdline.IsPresent(key) Then
                Return CInt(cmdline.GetValue(key))
            Else
                Return Nothing
            End If
        End Function
    End Module

End Namespace

File: Src/Icm.Core/CommandLine/CommandLineExtensions.cs

using Microsoft.VisualBasic;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Data;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using res = My.Resources.CommandLineTools;
using Icm.Collections;
using Icm.ColorConsole;

namespace Icm.CommandLineTools
{

	public static class CommandLineExtensions
	{

		/// <summary>
		///  Prints the set of instructions for a CommandLine.
		/// </summary>
		/// <remarks></remarks>
		[Extension()]
		public static void Instructions(CommandLine cmdline)
		{
			NamedParameter o = default(NamedParameter);

			Console.Write("{0}: ", res.S_HELP_USAGE, System.Diagnostics.Process.GetCurrentProcess.ProcessName);

			using (new ColorSetting(ConsoleColor.Cyan)) {
				Console.Write(System.Diagnostics.Process.GetCurrentProcess.ProcessName);

				foreach (void param_loopVariable in cmdline.NamedParameters) {
					param = param_loopVariable;
					if (param.IsRequired) {
						Console.Write(" --{0}", param.LongName);
					} else {
						Console.Write(" [--{0}", param.LongName);
					}
					foreach (void subarg_loopVariable in param.SubArguments) {
						subarg = subarg_loopVariable;
						if (subarg.IsRequired) {
							Console.Write(" " + subarg.Name);
						} else {
							Console.Write(" [{0}]", subarg.Name);
						}
					}
					if (!param.IsRequired) {
						Console.Write("]");
					}

				}
			}
			using (new ColorSetting(ConsoleColor.Green)) {

				var _with1 = cmdline.MainParameters;
				for (i = 1; i <= _with1.Minimum; i++) {
					Console.Write(" {0}", _with1.GetArgumentName(i).Name);
				}
				if (_with1.Maximum.HasValue) {
					for (i = _with1.Minimum + 1; i <= _with1.Maximum.V; i++) {
						Console.Write(" [{0}]", _with1.GetArgumentName(i).Name);
					}
				} else {
					Console.Write(" {{{0}}}", _with1.UnboundArgumentsName.Name);
				}
			}
			using (new ColorSetting(ConsoleColor.White)) {
				Console.WriteLine();
				Console.WriteLine();
				Console.Write("{0}:", res.S_OPTIONS);
				Console.WriteLine();

				foreach ( o in cmdline.NamedParameters) {
					List<string> subargs = new List<string>();

					foreach (SubArgument s in o.SubArguments) {
						if (s.IsRequired) {
							subargs.Add(s.Name);
						} else if (s.IsOptional) {
							subargs.Add(string.Format("[{0}]", s.Name));
						} else if (s.IsList) {
							subargs.Add(string.Format("{{{0}}}", s.Name));
						}
					}
					using (new ColorSetting(ConsoleColor.Cyan)) {
						Console.WriteLine("  -{0} {1}", o.ShortName, subargs.JoinStr(" "));
						Console.WriteLine("  --{0} {1}", o.LongName, subargs.JoinStr(" "));
					}
					Console.WriteLine("      {0}{1}", o.IsRequired ? res.S_HELP_REQUIRED : "", o.Description);
				}

				Console.WriteLine();
				var _with2 = cmdline.MainParameters;
				foreach (void key_loopVariable in _with2.ParamDictionary.Keys) {
					key = key_loopVariable;
					using (new ColorSetting(ConsoleColor.Green)) {
						Console.WriteLine("  {0}", key);
					}
					Console.WriteLine("      {0}", _with2.ParamDictionary(key).Description);
				}
				Console.WriteLine();
				Console.WriteLine(res.S_SUBARGUMENTS_SEPARATED);
			}

		}

		/// <summary>
		///  Prints the set of values for a CommandLine.
		/// </summary>
		/// <remarks></remarks>
		[Extension()]
		public static void Print(CommandLine cmdline)
		{
			using (new ColorSetting(ConsoleColor.Cyan)) {
				foreach (void param_loopVariable in cmdline.NamedParameters) {
					param = param_loopVariable;
					if (cmdline.IsPresent(param.ShortName)) {
						ColorConsole.ColorConsole.Write(ConsoleColor.Cyan, "{0}=", param.LongName);
						ColorConsole.ColorConsole.WriteLine(ConsoleColor.White, "({0})", cmdline.GetValues(param.ShortName).JoinStr("", "\"", ",", "\"", ""));
					} else {
						ColorConsole.ColorConsole.Write(ConsoleColor.Cyan, "{0}=", param.LongName);
						ColorConsole.ColorConsole.WriteLine(ConsoleColor.White, "UNDEFINED");
					}
				}
			}
			using (new ColorSetting(ConsoleColor.Green)) {

				Console.WriteLine(cmdline.MainValues.JoinStr("(", "\"", ",", "\"", ")"));
			}

			Console.WriteLine();

		}

		[Extension()]
		public static int? GetInteger(CommandLine cmdline, string key)
		{
			if (cmdline.IsPresent(key)) {
				return Convert.ToInt32(cmdline.GetValue(key));
			} else {
				return null;
			}
		}
	}

}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
